node/core/approval-voting/src/ops.rs

```diff
@@ -311,3 +311,76 @@ pub fn force_approve(
 
 	Ok(approved_hashes)
 }
+
+/// Revert to the block corresponding to the specified `hash`.
+/// The operation is not allowed for blocks older than the last finalized one.
+pub fn revert_to(
+	overlay: &mut OverlayedBackend<'_, impl Backend>,
+	hash: Hash,
+) -> SubsystemResult<()> {
+	let (children, height) = match overlay.load_block_entry(&hash)? {
+		Some(mut entry) => {
+			let children_height = entry.block_number() + 1;
+			let children = std::mem::take(&mut entry.children);
+			// Write revert point block entry without the children.
+			overlay.write_block_entry(entry);
+			(children, children_height)
+		},
+		None => {
+			let children_height =
+				overlay.load_stored_blocks()?.map(|range| range.0).ok_or_else(|| {
+					SubsystemError::Context(
+						"no available blocks to infer revert point height".to_string(),
+					)
+				})?;
+
+			let children = overlay.load_blocks_at_height(&children_height)?;
+
+			let child_entry = children
+				.first()
+				.and_then(|hash| overlay.load_block_entry(hash).ok())
+				.flatten()
+				.ok_or_else(|| {
+					SubsystemError::Context("lookup failure for first block".to_string())
+				})?;
+
+			// The parent is expected to be the revert point
+			if child_entry.parent_hash() != hash {
+				return Err(SubsystemError::Context(
+					"revert below last finalized block or corrupted storage".to_string(),
+				))
+			}
+
+			(children, children_height)
+		},
+	};
+
+	let mut stack: Vec<_> = children.into_iter().map(|h| (h, height)).collect();
+
+	while let Some((hash, number)) = stack.pop() {
+		let mut blocks_at_height = overlay.load_blocks_at_height(&number)?;
+		blocks_at_height.retain(|h| h != &hash);
+		overlay.write_blocks_at_height(number, blocks_at_height);
+
+		if let Some(entry) = overlay.load_block_entry(&hash)? {
+			overlay.delete_block_entry(&hash);
+
+			// Cleanup the candidate entries by removing any reference to the
+			// removed block. If for a candidate entry the block block_assignments
+			// drops to zero then we remove the entry.
+			for (_, candidate_hash) in entry.candidates() {
+				if let Some(mut candidate_entry) = overlay.load_candidate_entry(candidate_hash)? {
+					candidate_entry.block_assignments.remove(&hash);
+					if candidate_entry.block_assignments.is_empty() {
+						overlay.delete_candidate_entry(candidate_hash);
+					} else {
+						overlay.write_candidate_entry(candidate_entry);
+					}
+				}
+			}
+
+			stack.extend(entry.children.into_iter().map(|h| (h, number + 1)));
+		}
```

Comment:
This impl looks similar to chain-selection, which I haven't reviewed
But I think the logic to revert to block number N should be the following:
1. set stored range to N+1..=old_stored_range.1;
2. delete all blocks at height N+1..=old_stored_range.1;
3. delete all candidates from these blocks
4. delete these block entries.

Could you clarify why we're doing something slightly different here?

Reply:
Yeah, the logic in principle is quiet similar, what is different are the structures to remove from the database.
The two subsystems are using the database independently, so the data to remove is not the same.

E.g.:
- Even though both the subsystems store the block information in a structure called `BlockEntry`, the content (and the usage) is different.
- The approval-voting persists parachains blocks candidates, and these should be removed if are not referenced by any block
- The chain-selection subsystem persists what is the best candidate block information

Reply:
> Could you clarify why we're doing something slightly different here?

Sorry if that wasn't clear, by slightly different I meant not from chain-selection, but from what I expected to see as a reversion logic for approval-voting db.

A couple of more concrete questions:
1. Why don't we modify the stored range?
2. Why
```rust
	blocks_at_height.retain(|h| h != &hash);
	overlay.write_blocks_at_height(number, blocks_at_height);
```
instead of 
```
	overlay.delete_blocks_at_height(number)
```
Note that `blocks_at_heights` and `block_entry` are stored separately.

Reply:
> 0. Sorry if that wasn't clear, by slightly different I meant not from chain-selection, but from what I expected to see as a reversion logic for approval-voting db.

We dong want to delete all blocks with height greater than the revert point. But only the ones that are children on the revert point.
There could be blocks with height greater that the revert point BUT that are forking from below. We don't want to remove such blocks.

> 1. Why don't we modify the stored range?

Good catch. That needs to be written as well. Fix in progress...

> 2. Why ...

This is because of point 0. We don't want to remove every block with height greater than the revert point. So what we do is we remove the current one (a descendant of the revert point) but we retain the ones with the same height that are not descendants

I hope this clarifies

Reply:
> We don't want to delete all blocks with height greater than the revert point. But only the ones that are children on the revert point. There could be blocks with height greater that the revert point BUT that are forking from below it. We don't want to remove such blocks.

Thanks for the clarification.
So we're intentionally pruning a subtree of some canonical block at a given height, but not any other forks.
Intuitively, it's not clear to me why this is behavior is what a user expects from this subcommand. But if it is, the impl seems correct. Thanks for bearing with me :)

Reply:
Thank you for spotting a bug and an optimization ;)